dask_cuda/explicit_comms/dataframe_shuffle.py

```diff
@@ -0,0 +1,260 @@
+    trim_map = {}
+    for i in range(s["nworkers"]):
+        if i in workers:
+            trim_map[i] = len(trim_map)
+
+    rank = trim_map[s["rank"]]
+    eps = {trim_map[i]: s["eps"][trim_map[i]] for i in workers if i != s["rank"]}
+
+    df = df_concat(df_parts)
+
+    return await rearrange_by_column(len(workers), rank, eps, df, column)
+
+
+def dataframe_rearrange_by_column(df, column):
+    """Order divisions of DataFrame so that all values within column(s) align
+
+    This enacts a task-based shuffle.  It contains most of the tricky logic
+    around the complex network of tasks.  Typically before this function is
+    called a new column, ``"_partitions"`` has been added to the dataframe,
+    containing the output partition number of every row.  This function
+    produces a new dataframe where every row is in the proper partition.  It
+    accomplishes this by splitting each input partition into several pieces,
+    and then concatenating pieces from different input partitions into output
+    partitions.
+
+    Note that the `column` input may correspond to a list of columns (rather
+    than just a single column name).  In this case, the `shuffle_group` and
+    `shuffle_group_2` functions will use hashing to map each row to an output
+    partition. This approach may require the same rows to be hashed multiple
+    times, but avoids the need to assign a new "_partitions" column.
+
+    Parameters
+    ----------
+    df: dask.dataframe.DataFrame
+    column: str or list
+        A column name on which we want to split, commonly ``"_partitions"``
+        which is assigned by functions upstream.  This could also be a list of
+        columns (in which case shuffle_group will create a hash array/column).
+
+    Returns
+    -------
+    df3: dask.dataframe.DataFrame
+
+    See also
+    --------
+    rearrange_by_column_disk: same operation, but uses partd
+    rearrange_by_column: parent function that calls this or rearrange_by_column_disk
+    shuffle_group: does the actual splitting per-partition
+    """
+
+    return comms.default_comms().dataframe_operation(
+        _rearrange_by_column,
+        df_list=(df,),
+        extra_args=(column,),
+    )
+
+
+def dataframe_shuffle(df, index):
+    """Group DataFrame by index
+
+    Hash grouping of elements. After this operation all elements that have
+    the same index will be in the same partition. Note that this requires
+    full dataset read, serialization and shuffle. This is expensive. If
+    possible you should avoid shuffles.
+
+    This does not preserve a meaningful index/partitioning scheme. This is not
+    deterministic if done in parallel.
+
+    As a side effect, this operation concatenate all partitions located on
+    the same worker thus npartitions of the returned dataframe equals number
+    of workers.
+
+    See Also
+    --------
+    set_index
+    set_partition
+    shuffle_disk
+    """
+
+    if isinstance(index, str) or (
```

Comment:
It looks like this is the only code path with test coverage, and only for the `str` case.  We should add a test for a list of `str` column names, and maybe for a collection-based index.